### DISCUSSION TITLE
problemas con conexion python a tia portal

### DISCUSSION THREAD

**Discussion by ignaciofigue** [gijzelaerr/python-snap7#491] [Q&A] [answered]
I'm having issues with the following code in the part where it says datos = simular_datos_balanza. I added int() to convert the value to a numeric one, but it raises an error when reaching that line (int() argument must be a string, a bytes-like object, or a real number, not 'function').

I'm posting my code here below to see if someone can help me. I'm trying to simulate sending a weight to the general server when passing my hand over a sensor. Thank you in advance.

import socket
import random
import time
import snap7

def simular_datos_balanza():
    # Simula la generación de datos de la balanza
    peso = random.uniform(-0.1, 5.0)  # Genera un peso aleatorio entre -0.1 y 5.0 kg
    return f"{peso:.2f} kg"

def recibir_pulso_sensor(plc):
    data = plc.read_area(0x82, 0, 0, 1)
    return data[0] == 1

def iniciar_cliente(host, puerto):
    try:
        # Configurar el socket del cliente
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as cliente_socket:
            cliente_socket.connect((host, puerto))
            print(f"Conexión establecida con el servidor en {host}:{puerto}")

            # Configurar la conexión con el PLC (TIA Portal)
            plc = snap7.client.Client()
            plc.connect('192.168.0.2', 0, 1)  # Reemplaza con la dirección IP de tu PLC

            while True:
                # Simular datos de la balanza
                print("conexion establecida con el plc")
                time.sleep(3)

                datos = int(simular_datos_balanza)
                # datos = simular_datos_balanza()

                # Verificar el pulso del sensor en TIA Portal
                pulso_sensor =recibir_pulso_sensor(plc)

                # Convertir el peso a un número para aplicar el filtro
                peso = (datos.split()[0])

                # Enviar datos al servidor solo si se recibe un pulso del sensor
                if pulso_sensor:
                    cliente_socket.sendall(datos.encode('utf-8'))
                    print(f"Datos enviados al servidor: {datos}")
                    # Esperar el intervalo antes de enviar el siguiente conjunto de datos
                    
            else:
             print("Esperando pulso del sensor...")

    except Exception as e:
        print(f"Error de conexión: {e}")
    finally:
        # Cerrar la conexión con el PLC al finalizar
        plc.disconnect()

# # Configuración del servidor, puerto y intervalos de tiempo
servidor_host = '192.168.0.2'  # Cambia esto a la dirección del servidor
servidor_puerto = 2147  # Cambia esto al puerto del servidor
 

# # Llamar a la función para iniciar el cliente
iniciar_cliente(servidor_host, servidor_puerto)

**Comment by lupaulus** [comment]
@ignaciofigue I don't understand the half of your post because of mixing English and Spainish with your code snippets. Can you reformat the post ?
TIA Portal is not a Siemens CPU but the software for developing on CPU. What is your CPU or do you use PLC SIM ?

**Comment by lupaulus** [comment]
@ignaciofigue 
```
def simular_datos_balanza():
# Simula la generación de datos de la balanza
peso = random.uniform(-0.1, 5.0) # Genera un peso aleatorio entre -0.1 y 5.0 kg
return f"{peso:.2f} kg" # THIS DOESN'T RETURN an INT ...
```

this DOESN'T return an int ...

`return f"{peso:.2f} kg" ` 

Your Issue isn't related to your library it will be closed

**Comment by gijzelaerr** [comment, marked as answer]
Closing as resolved. This was a basic Python coding error (missing parentheses in function call), not a python-snap7 issue.
